### PR TITLE
Added embeddings support for ollama

### PR DESCRIPTION
**Pull request by yackermann**
cc  https://github.com/jmorganca/ollama/issues/436

### CODE REVIEW

**Comment by herrjemand** [comment]
TODO:
- [ ] Tests
- [ ] Python workbook

**Comment by baskaryan** [comment]
cc @jmorganca

File: libs/langchain/langchain/embeddings/ollama.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    """Instruction used to embed the query."""
+    model_kwargs: Optional[dict] = None
+    """Other model keyword args"""
+
+    class Config:
+        """Configuration for this pydantic object."""
+        extra = Extra.forbid
+
+    @root_validator()
+    def validate_environment(cls, values: Dict) -> Dict:
```
Comment (leo-gan) [inline libs/langchain/langchain/embeddings/ollama.py:47]: It seems this function does nothing. Could it be removed?

File: libs/langchain/langchain/embeddings/ollama.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+        Args:
+            text: The text to embed.
+
+        Returns:
+            Embeddings for the text.
+        """
+        instruction_pair = f"{self.query_instruction}{text}"
+        embedding = self._embed([instruction_pair])[0]
+        return embedding
```
Comment (leo-gan) [inline libs/langchain/langchain/embeddings/ollama.py:123]: @herrjemand Thanks for a great integration!
Would you like to create an example for it in a separate PR or would you like for someone else to make it? Please, let us know.

Comment (hwchase17) [reply libs/langchain/langchain/embeddings/ollama.py:123]: agree, this is fantastic! lets def get an example notebook in

File: libs/langchain/langchain/embeddings/ollama.py (hunk 1 above)
Comment (herrjemand) [reply libs/langchain/langchain/embeddings/ollama.py]: Addressed

File: libs/langchain/langchain/embeddings/ollama.py (hunk 2 above)
Comment (herrjemand) [reply libs/langchain/langchain/embeddings/ollama.py:116]: Addressed *)

**Comment by Basti-an** [comment]
FYI: 

The api/embedding route of the ollama server supports the same "advanced options" as the api/generate route (although this is not clearly documented by Ollama). 

See [docs](https://github.com/jmorganca/ollama/blob/main/docs/modelfile.md#valid-parameters-and-values) and [embedding handler code](https://github.com/jmorganca/ollama/blob/main/server/routes.go#L242C45-L242C56).

Some of these options, such as "use_mmap" can improve embedding times dramatically, so I would recommend adding support for these options.

**Comment by herrjemand** [comment]
Thanks @Basti-an. Added fix.

Additionally I've added fix to the ollama docs:

Ref: https://github.com/jmorganca/ollama/pull/472

File: libs/langchain/langchain/embeddings/ollama.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    def _embed(self, input: List[str]) -> List[List[float]]:
+        embeddings_list: List[List[float]] = []
+        for prompt in input:
+            embeddings = self._process_emb_response(prompt)
+            embeddings_list.append(embeddings)
+
+        return embeddings_list
+
+    def embed_documents(self, texts: List[str]) -> List[List[float]]:
+        """Embed documents using a Deep Infra deployed embedding model.
```
Comment (jmorganca) [inline libs/langchain/langchain/embeddings/ollama.py:182]: Should this refer to `Ollama`?

File: libs/langchain/langchain/embeddings/ollama.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+        Returns:
+            List of embeddings, one for each text.
+        """
+        instruction_pairs = [f"{self.embed_instruction}{text}" for text in texts]
+        embeddings = self._embed(instruction_pairs)
+        return embeddings
+
+    def embed_query(self, text: str) -> List[float]:
+        """Embed a query using a Deep Infra deployed embedding model.
```
Comment (jmorganca) [inline libs/langchain/langchain/embeddings/ollama.py:195]: Same as above comment :-)

**Comment by jmorganca** [comment]
Overall looks great! Nice work adding this! Just a small note on some of the comments

**Comment by herrjemand** [comment]
I think its ready to merge

**Comment by baskaryan** [comment]
awesome stuff @herrjemand! can we add a quick notebook to `docs/extras/integrations/text_embedding/` (instead of adding to the LLM one)?

**Comment by herrjemand** [comment]
@baskaryan done! *)

**Pull request merged**